File: ruby/easy/minimum_absolute_difference_in_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def get_minimum_difference(root)
    hash = { min: nil, prev: nil }
    min_abs_distance(root, hash)
    return hash[:min] ? hash[:min] : 0
end

def min_abs_distance(root, hash)
    return unless root
    min_abs_distance(root.left, hash)
    if hash[:prev]
        min = hash[:min]
        diff = (root.val - hash[:prev]).abs
        hash[:min] = min ? [min, diff].min : diff
    end
    hash[:prev] = root.val
    min_abs_distance(root.right, hash)
end

File: javascript/easy/minimumAbsoluteDifferenceInBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {Integer}
 */
var getMinimumDifference = function(root) {
    const obj = { min: null, prev: null };
    getMinimumAbsDistance(root, obj);
    return obj.min === null ? 0 : obj.min;
};

var getMinimumAbsDistance = function(root, obj) {
    if(root === null) return;
    getMinimumAbsDistance(root.left, obj);
    if(obj.prev !== null) {
        let min = obj.min, diff = Math.abs(root.val - obj.prev);
        obj.min = ((min === null) ? diff : Math.min(min, diff));
    }
    obj.prev = root.val;
    getMinimumAbsDistance(root.right, obj);
};
